lectures/sympy.md

```diff
@@ -0,0 +1,861 @@
+expr_sub
+```
+
+```{code-cell} ipython3
+solve(Eq(expr_sub, 1))
+```
+
++++ {"user_expressions": []}
+
+Below is another example equation with symbol `x` and functions `sin`, `cos`, and `tan` using the `Eq` function
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 17.5%
+---
+# Create an equation
+eq = Eq(cos(x) / (tan(x)/sin(x)), 0)
+eq
+```
+
++++ {"user_expressions": []}
+
+Now we simplify this equation using the `simplify` function
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 10%
+---
+# Simplify an expression
+simplified_expr = simplify(eq)
+simplified_expr
+```
+
++++ {"user_expressions": []}
+
+We can solve equations using the `solve` function in SymPy
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 5.5%
+---
+# Solve the equation
+sol = solve(eq, x)
+sol
+```
+
++++ {"user_expressions": []}
+
+SymPy can also handle more complex equations involving trigonometry and complex numbers.
+
+We demonstrate this using [Euler's formula](https://en.wikipedia.org/wiki/Euler%27s_formula)
+
+```{code-cell} ipython3
+# 'I' represents the imaginary number i 
+euler = cos(x) + I*sin(x)
+euler
+```
+
+```{code-cell} ipython3
+simplify(euler)
+```
+
++++ {"user_expressions": []}
+
+#### Example: fixed point computation
+
+Fixed point computation is a common problem in economics and finance.
+
+Here we solve the fixed point of the Solow-Swan growth dynamics:
+
+$$
+k_{t+1}=s f\left(k_t\right)+(1-\delta) k_t, \quad t=0,1, \ldots
+$$
+
+where $k_t$ is the capital stock, $f$ is a production function, $\delta$ is a rate of depreciation.
+
+With $f(k) = Ak^a$, we can show the unique fixed point of the dynamics using pen and paper:
+
+$$
+k^*:=\left(\frac{s A}{\delta}\right)^{1 /(1-\alpha)}
+$$ 
+
+This can be easily computed in SymPy
+
+```{code-cell} ipython3
+A, s, k, α, δ = symbols('A s k α δ')
+```
+
+```{code-cell} ipython3
+# Define Solow-Swan growth dynamics
+solow = Eq(s*A*k**α + (1 - δ) * k, k)
+solow
+```
+
+```{code-cell} ipython3
+solve(solow, k)
+```
+
+### Inequalities and logic
+
+SymPy also allows users to define inequalities and set operators and provides a wide range of [operations](https://docs.sympy.org/latest/modules/solvers/inequalities.html).
+
+```{code-cell} ipython3
+reduce_inequalities([2*x + 5*y <= 30, 4*x + 2*y <= 20], [x])
+```
+
+```{code-cell} ipython3
+And(2*x + 5*y <= 30, x > 0)
+```
+
++++ {"user_expressions": []}
+
+### Series
+
+Series are widely used in economics and statistics, from asset pricing to the expectation of discrete random variables.
+
+We can construct a simple series of summations using `Sum` function
+
+```{code-cell} ipython3
+x, y, i, j = symbols("x y i j")
+sum_xy = Sum(Indexed('x',i) * Indexed('y', j), 
+            (i, 0, 3),
+            (j, 0, 3))
+sum_xy
+```
+
++++ {"user_expressions": []}
+
+To evaluate the sum, we can `lamdify` the formula.
+
+The lamdified expression can take vectors as input for $x$ and $y$ and compute the result
```

Comment:
```suggestion
The lamdified expression can take numeric values as input for $x$ and $y$ and compute the result
```

Lambdify in general works for all numeric inputs. See: https://docs.sympy.org/latest/modules/utilities/lambdify.html#sympy.utilities.lambdify.lambdify